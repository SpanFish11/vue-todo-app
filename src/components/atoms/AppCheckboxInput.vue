<script setup>
const props = defineProps({
  disabled: { type: Boolean, required: false, default: false },
  checked: { type: Boolean, required: false, default: false },
})

const emit = defineEmits(['onChange'])
</script>
`
<template>
  <input
    class="form-check-input"
    type="checkbox"
    :disabled="props.disabled"
    :checked="props.checked"
    @change="emit('onChange')"
  />
</template>

<style scoped lang="scss">
.form-check-input {
  border: transparent;
  width: 23px;
  height: 23px;
  background: #c0c0c0;
  -webkit-mask: url('../../assets/icons/circle.svg') no-repeat center;
  -webkit-mask-size: contain;
  mask: url('../../assets/icons/circle.svg') no-repeat center;
  mask-size: contain;
  stroke: none;

  &:focus {
    box-shadow: none;
  }

  &:checked {
    border: none;
    width: 23px;
    height: 23px;
    background: linear-gradient(130deg, rgba(25, 134, 217, 0.65), rgba(200, 57, 222, 0.72));
    -webkit-mask: url('../../assets/icons/check-circle-fill.svg') no-repeat center;
    -webkit-mask-size: contain;
    mask: url('../../assets/icons/check-circle-fill.svg') no-repeat center;
    mask-size: contain;
  }

  &:hover {
    cursor: pointer;
    background: linear-gradient(130deg, rgba(38, 121, 180, 0.6526260162268032), rgba(189, 58, 208, 0.7198529069831057));
  }
}
</style>
