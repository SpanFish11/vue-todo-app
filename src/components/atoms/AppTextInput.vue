<script setup>
const props = defineProps({
  placeholder: { type: String, required: false, default: 'Add your task' },
})
</script>

<template>
  <input type="text" class="text-input" :placeholder="props.placeholder" />
</template>

<style scoped lang="scss">
.text-input {
  width: 100%;
  padding: 10px;
  margin-bottom: 20px;
  border-radius: 5px;
  border: 1px solid #ddd;
}
</style>
