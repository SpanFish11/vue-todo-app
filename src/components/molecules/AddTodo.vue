<script setup>
import AppTextInput from '@/components/atoms/AppTextInput.vue'
import AppCheckboxInput from '@/components/atoms/AppCheckboxInput.vue'
import { ref } from 'vue'
import { useStore } from 'vuex'

const store = useStore()
const inputValue = ref('')

const addTodo = (value) => store.dispatch('addTodo', { value })

const doneEdit = () => {
  const value = inputValue.value.trim()
  if (value) {
    addTodo(value)
    inputValue.value = ''
  }
}

const cancelEdit = () => (inputValue.value = '')
</script>

<template>
  <div class="card mb-4">
    <div class="card-body">
      <div class="input-group">
        <div class="input-group-text">
          <AppCheckboxInput disabled />
        </div>
        <AppTextInput
          class="form-control"
          v-model="inputValue"
          placeholder="Add your todo"
          :maxlength="150"
          @onPressEnter="doneEdit"
          @onPressEsc="cancelEdit"
        />
      </div>
    </div>
  </div>
</template>

<style scoped lang="scss">
.card-body {
  padding: 10px;
  margin-left: 10px;
}

.form-control {
  border: none;

  &:focus {
    box-shadow: none;
  }
}

.input-group-text {
  border: none;
  background-color: white;
  padding-left: 0;
}
</style>
