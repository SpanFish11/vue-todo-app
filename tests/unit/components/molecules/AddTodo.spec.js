import { beforeEach, describe, expect, it } from '@jest/globals'
import { shallowMount, mount } from '@vue/test-utils'
import AddTodo from '@/components/molecules/AddTodo.vue'
import { createStore } from 'vuex'
import AppCheckboxInput from '@/components/atoms/AppCheckboxInput.vue'
import AppTextInput from '@/components/atoms/AppTextInput.vue'

describe('Given the AddTodo component', () => {
  describe('when it is rendered', () => {
    let wrapper

    beforeEach(() => {
      wrapper = shallowMount(AddTodo, {})
    })

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it('should render InputCheckbox with default props', () => {
      const appCheckboxInput = wrapper.findComponent(AppCheckboxInput)

      expect(appCheckboxInput.exists()).toBe(true)
      expect(appCheckboxInput.props('disabled')).toBe(true)
      expect(appCheckboxInput.props('checked')).toBe(false)
    })

    it('should render TextInput with default props', () => {
      const appTextInput = wrapper.findComponent(AppTextInput)

      expect(appTextInput.exists()).toBe(true)
      expect(appTextInput.props('placeholder')).toBe('Add your todo')
      expect(appTextInput.props('maxlength')).toBe(150)
    })
  })

  describe('when interacting with AppTextInput', () => {
    let store
    let wrapper

    beforeEach(() => {
      store = createStore({})
      store.dispatch = jest.fn()

      wrapper = mount(AddTodo, {
        global: {
          plugins: [store],
        },
      })
    })

    it('should dispatch and clear the input when Enter is pressed', async () => {
      const appTextInput = wrapper.findComponent(AppTextInput)

      expect(appTextInput.props().modelValue).toBe('')

      await appTextInput.setValue(' test ')
      await appTextInput.trigger('keyup.enter')
      await wrapper.vm.$nextTick()

      expect(store.dispatch).toHaveBeenCalledWith('addTodo', { value: 'test' })
      expect(appTextInput.props().modelValue).toBe('')
    })

    it('should do nothing when the input is clear and Enter is pressed', async () => {
      const appTextInput = wrapper.findComponent(AppTextInput)

      await appTextInput.trigger('keyup.enter')
      await wrapper.vm.$nextTick()

      expect(store.dispatch).not.toBeCalled()
      expect(appTextInput.props().modelValue).toBe('')
    })

    it('should clear the input when Escape is pressed', async () => {
      const appTextInput = wrapper.findComponent(AppTextInput)

      await appTextInput.setValue(' test ')

      expect(appTextInput.props().modelValue).toBe(' test ')

      await appTextInput.trigger('keyup.esc')
      await wrapper.vm.$nextTick()

      expect(store.dispatch).not.toBeCalled()
      expect(appTextInput.props().modelValue).toBe('')
    })
  })
})
