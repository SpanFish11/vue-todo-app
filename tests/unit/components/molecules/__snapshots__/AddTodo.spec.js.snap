// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the AddTodo component when it is rendered should match the snapshot 1`] = `
<div
  class="card mb-4"
>
  <div
    class="card-body"
  >
    <div
      class="input-group"
    >
      <div
        class="input-group-text"
      >
        <app-checkbox-input-stub
          checked="false"
          disabled="true"
        />
      </div>
      <app-text-input-stub
        class="form-control"
        maxlength="150"
        modelvalue=""
        placeholder="Add your todo"
      />
    </div>
  </div>
</div>
`;
